Grid Renderer - Improved appearance of grids

bubblobill opened the pull request:
### Identify the Bug or Feature request
progresses #3840, #167
resolves #3674

### Description of the Change
Finally realises the intent of my change for GridLineWidth to GridLineWeight.

Separates the rendering of the grid from the construction of the grid shape(s) to draw.
Square, Iso, and Hex Grid create the shape to draw and pass it back to GridRenderer for actual painting.
Square Grid - changed coordinates to have drop shadow on all sides to improve legibility.

GridRenderer 
- scales drawing stroke according to zone scale.
- bases stroke on size of grid (2% of size) for consistent appearance between grid sizes.
- paints grids with a translucent contrast shadow above 50% zoom.

Did not touch the LibGDX stuff other than a refactor.

### Possible Drawbacks
Should be none.

### Documentation Notes
Improved appearance of grids

### Release Notes

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/5581)
<!-- Reviewable:end -->

bubblobill committed:
GridRenderer - takes back the actual drawing of grid shapes.
- scales drawing stroke according to zone scale.
- bases stroke on size of grid for consistent appearance (2% of size) between grid
sizes.
- paints grids with a translucent contrast shadow above 50% zoom.
Square grid
- changed coordinates to have drop shadow on all sides to
improve legibility.
Square, Iso, and Hex Grid - pass rendering back to grid renderer for
actual drawing.

bubblobill committed:
change gridSize to more meaningful geidLineWeight

bubblobill committed:
change gridSize to more meaningful geidLineWeight

bubblobill committed:
Merge branch 'develop' into grid-renderer

bubblobill committed:
change gridSize to more meaningful geidLineWeight

bubblobill committed:
change gridSize to more meaningful gridLineWeight

bubblobill committed:
removed reference to gridLineWeight

bubblobill committed:
I hate spot

bubblobill committed:
Merge branch 'RPTools:develop' into grid-renderer

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:
```diff
@@ -14,31 +14,93 @@
+  private static float gridLineWeight;
+  private static float scale;
+  private static float baseWidth = 2f;
+  private static Color[] gridColours;
```
When the current grid changes (either by modifying a map or loading a different map), the `gridColours` needs to be recalculated.

bubblobill replied:
Sure, and the GridChanged event actually works. </sarcasm>

kwvanderlinde replied:
Creating the `gridColours` array is not at all expensive. You might as well just do it every time in `renderGrid()` and don't worry about caching it.

bubblobill replied:
Oh no. Can't do that.
Step 2. Calculate grid colours based on type of display. Line - dashed - dotted - multiply. Calculate negative colour using perception based calculations. Interpolate colours using easing functions. 
2 a) Make considerable more than 4 painting passes, or 
2 b) Create an image to paint instead, or
2 c) Find exotic things on the internet to try.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:
```diff
@@ -14,31 +14,93 @@
+        || !AppState.isShowGrid()
+        || zone.getGrid().getSize() * renderer.getScale() < ZoneRendererConstants.MIN_GRID_SIZE) {
+      if (renderer != null) {
+        setRenderer(renderer);
```
I don't think this call does anything.

bubblobill replied:
Sure it does. If there is no Zone, set the renderer again. Rinse and repeat.

kwvanderlinde replied:
Ugh, this is the problem with handling nulls in such a dynamic way.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:
```diff
@@ -14,31 +14,93 @@
+
+  public void setRenderer(ZoneRenderer renderer) {
     this.renderer = renderer;
+    if (renderer.getZone() == null) {
```
Renderers don't have null zones.

bubblobill replied:
Sure they do. Take it out and discover a whole new world.

bubblobill replied:
No zone during ZR  construction.

kwvanderlinde replied:
```java
  public ZoneRenderer(Zone zone) {
    if (zone == null) {
      throw new IllegalArgumentException("Zone cannot be null");
    }
    this.zone = zone;
```
They shouldn't, not even during construction.

Regardless, you can also avoid it by passing the `Zone` to the constructor like many other renderers do and you can dispense with this and other null checks for the zone.

bubblobill replied:
This is a minimal-change/initial steps PR. Keeping the previous bullet-proofing and stuff.
I will probably institute your RenderBuddy class down the road. For now, all these checks and stuff, make sure that ZRs without Zones, and Zones without Grids, etc. don't cause screwups. Also, it fails tests if I remove the zone null check.

kwvanderlinde replied:
Just checked the tests, they work fine without the null handling.

bubblobill replied:
Well they were failing on something somewhere until I spammed null handling everywhere.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:
What do you think about moving the various `Grid#draw(...)` implementations into `GridRenderer`? It would avoid the need to make half the class static and should simplify some of the state management.

bubblobill replied:
There's a reason this is step one on the journey to prettier grids.

This gets gridRenderer actually painting the various grids.

I intend to construct the grids in a consistent fashion later whilst adding fruit.

bubblobill committed:
Merge branch 'develop' into grid-renderer

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:
```diff
@@ -14,31 +14,93 @@
+    setRenderer(renderer);
+  }
+
+  public void setRenderer(ZoneRenderer renderer) {
```
Method should be `private`.

bubblobill replied:
Nope. Because GridChanged is broken, this will be called in ZR onGridChanged. In fact, I'll do that now and add colour calculation to setRenderer().

kwvanderlinde replied:
In that case, do something a little different: set `renderer` directly in the constructor and leave it at that. Then have a different method (`gridChanged()` or whatever you want to call it) that can udpate the the `GridRenderer` state based on the `renderer` that you already have.

That makes it a lot easier to follow the intent compared to possibly having arbitrary `ZoneRenderer` being injected at any time.

bubblobill replied:
changes to grid colour via the grid edit tool should trigger gridchanged event, which triggers ZR to update renderer on grid dependant renderers.

kwvanderlinde replied:
Please re-read what I wrote.

You don't need to support setting some arbitrary `ZoneRenderer` since `GridRenderer` _already has the `ZoneRenderer`._ This would suffice instead:
```java
  public void onGridChanged() {
    this.zone = renderer.getZone();
    setGridColours();
  }
```

bubblobill replied:
Then it will have to wait until not 2:45am.

kwvanderlinde replied:
Oh yeah, get some sleep :sweat_smile:

bubblobill replied:
sorted

kwvanderlinde commented:
As long as the grid colour stuff is made to work, I'm happy with the PR. I'd encourage you to double-check the need for the null checks, but I'm not going to push that too hard.

bubblobill committed:
Calculate colour array as part of setRenderer.
Set renderer from ZR onGridChanged.

bubblobill committed:
Calculate colour array as part of setRenderer.
Set renderer from ZR onGridChanged.

bubblobill committed:
Merge branch 'develop' into grid-renderer

bubblobill committed:
Deferred gridChange event firing to prevent firing with null zone during
 construction.

bubblobill committed:
Removed gridRenderer.setRenderer().
Removed renderer and zone checks from renderGrid().
Added check for colour change.
Added GridChange event listener to GridRenderer.

bubblobill committed:
undid change to Zone

bubblobill committed:
Merge branch 'RPTools:develop' into grid-renderer
